activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/cmd/CreateProcessInstanceCmd.java

```diff
@@ -0,0 +1,100 @@
+import org.activiti.engine.repository.ProcessDefinition;
+import org.activiti.engine.runtime.ProcessInstance;
+
+import java.util.Map;
+
+public class CreateProcessInstanceCmd implements Command<ProcessInstance> {
+
+    private static final long serialVersionUID = 1L;
+    protected String processDefinitionKey;
+    protected String processDefinitionId;
+    protected Map<String, Object> variables;
+    protected Map<String, Object> transientVariables;
+    protected String businessKey;
+    protected String tenantId;
+    protected String processInstanceName;
+    protected ProcessInstanceHelper processInstanceHelper;
+
+    public CreateProcessInstanceCmd(String processDefinitionKey, String processDefinitionId, String businessKey, Map<String, Object> variables) {
+        this.processDefinitionKey = processDefinitionKey;
+        this.processDefinitionId = processDefinitionId;
+        this.businessKey = businessKey;
+        this.variables = variables;
+    }
+
+    public CreateProcessInstanceCmd(String processDefinitionKey, String processDefinitionId, String businessKey, Map<String, Object> variables, String tenantId) {
+        this(processDefinitionKey, processDefinitionId, businessKey, variables);
+        this.tenantId = tenantId;
+    }
+
+    public CreateProcessInstanceCmd(ProcessInstanceBuilderImpl processInstanceBuilder) {
+        this(processInstanceBuilder.getProcessDefinitionKey(),
+            processInstanceBuilder.getProcessDefinitionId(),
+            processInstanceBuilder.getBusinessKey(),
+            processInstanceBuilder.getVariables(),
+            processInstanceBuilder.getTenantId());
+        this.processInstanceName = processInstanceBuilder.getProcessInstanceName();
+        this.transientVariables = processInstanceBuilder.getTransientVariables();
+    }
+
+    public ProcessInstance execute(CommandContext commandContext) {
+
+        if (processDefinitionId == null && processDefinitionKey == null) {
+            throw new ActivitiIllegalArgumentException("processDefinitionKey and processDefinitionId are null");
+        }
+
+        DeploymentManager deploymentCache = commandContext.getProcessEngineConfiguration().getDeploymentManager();
+        ProcessDefinition processDefinition = this.getProcessDefinitionByProcessDefinitionId(processDefinitionId, deploymentCache);
+        if(processDefinition == null) {
+            processDefinition = (processDefinitionKey != null && hasNoTenant(tenantId)) ?
+                this.getProcessDefinitionByProcessDefinitionKey(processDefinitionKey, tenantId, deploymentCache):
+                this.getProcessDefinitionByProcessDefinitionKeyAndTenantId(processDefinitionKey, tenantId, deploymentCache);
+            if (processDefinition == null) {
+                throw new ActivitiObjectNotFoundException("No process definition found for key '" + processDefinitionKey + "' for tenant identifier " + tenantId, ProcessDefinition.class);
+            }
+        }
```

Comment:
I'd prefer to have `CreateProcessInstanceCmd` and `StartProcessInstanceCmd` sharing the same code to retrieve the process definition.

Reply:
I have created a separated task for this, and will do this when picking up that one.